src/runtime/rocm/rocm_module.cc

```diff
@@ -0,0 +1,278 @@
+    // must recheck under the lock scope
+    if (module_[device_id] == nullptr) {
+      ROCM_DRIVER_CALL(hipModuleLoadData(&(module_[device_id]), data_.c_str()));
+    }
+    hipFunction_t func;
+    hipError_t result = hipModuleGetFunction(&func, module_[device_id], func_name.c_str());
+    if (result != hipSuccess) {
+      LOG(FATAL)
+          << "ROCMError: hipModuleGetFunction " << func_name
+          << " failed with error: " << hipGetErrorString(result);
+    }
+    return func;
+  }
+  // get a global var from primary context in device_id
+  hipDeviceptr_t GetGlobal(int device_id,
+                        const std::string& global_name,
+                        size_t expect_nbytes) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    // must recheck under the lock scope
+    if (module_[device_id] == nullptr) {
+      ROCM_DRIVER_CALL(hipModuleLoadData(&(module_[device_id]), data_.c_str()));
+    }
+    hipDeviceptr_t global = nullptr;
+    size_t nbytes = 0;
+
+    hipError_t result = hipSuccess;
+    // ROCM doesn't support hipModuleGetGlobal yet.
+    // hipError_t result = hipModuleGetGlobal(&global, &nbytes,
+    //                                    module_[device_id], global_name.c_str());
+    CHECK_EQ(nbytes, expect_nbytes);
+    if (result != hipSuccess) {
+      LOG(FATAL)
+          << "ROCMError: hipModuleGetGlobal " << global_name
+          << " failed with error: " << hipGetErrorString(result);
+    }
+    return global;
+  }
+
+ private:
+  // the binary data
+  std::string data_;
+  // The format
+  std::string fmt_;
+  // function information table.
+  std::unordered_map<std::string, FunctionInfo> fmap_;
+  // The hip source.
+  std::string hip_source_;
+  // the internal modules per GPU, to be lazily initialized.
+  std::array<hipModule_t, kMaxNumGPUs> module_;
+  // internal mutex when updating the module
+  std::mutex mutex_;
+};
+
+// a wrapped function class to get packed fucn.
+class ROCMWrappedFunc {
+ public:
+  // initialize the ROCM function.
+  void Init(ROCMModuleNode* m,
+            std::shared_ptr<ModuleNode> sptr,
+            const std::string& func_name,
+            size_t num_void_args,
+            const std::vector<std::string>& thread_axis_tags) {
+    m_ = m;
+    sptr_ = sptr;
+    func_name_ = func_name;
+    std::fill(fcache_.begin(), fcache_.end(), nullptr);
+    thread_axis_cfg_.Init(num_void_args, thread_axis_tags);
+  }
+  // invoke the function with void arguments
+  void operator()(TVMArgs args,
+                  TVMRetValue* rv,
+                  void** void_args) const {
+    int device_id;
+    ROCM_CALL(hipGetDevice(&device_id));
+    if (fcache_[device_id] == nullptr) {
+      fcache_[device_id] = m_->GetFunc(device_id, func_name_);
+    }
+    hipStream_t strm = static_cast<hipStream_t>(ROCMThreadEntry::ThreadLocal()->stream);
+    ThreadWorkLoad wl = thread_axis_cfg_.Extract(args);
+    // HIP supports only extra_args.
+    ROCM_DRIVER_CALL(hipModuleLaunchKernel(
```

Comment:
Just to confirm, this relies on the fact that hipModuleLaunchKernel is thread-safe as in CUDA driver, please confirm if that is the case.